### PR TITLE
[BEAM-10708] Clean up pipeline instrument

### PR DESCRIPTION
**Pull request by KevinGG**
1. Moved cacheable and cachekey to dataclasses.
2. Cleaned up logic of finding the user defined pipeline.
3. Fixed missing add_derived_pipeline in background_caching_job module.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by KevinGG** [comment]
R: @TheNeuralBit 
R: @rohdesamuel 

PTAL, thx!

**Comment by rohdesamuel** [comment]
lgtm

**Review by rohdesamuel — APPROVED** [review]

File: sdks/python/apache_beam/runners/interactive/caching/cacheable.py
Comment (TheNeuralBit) [inline sdks/python/apache_beam/runners/interactive/caching/cacheable.py:48]: It [looks like](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) you can add `@dataclass(frozen=True)` and it will generate a `__hash__` implementation for you. The trade off is that assigning to fields generates an exception. Would that work for you?

File: sdks/python/apache_beam/runners/interactive/pipeline_instrument.py
Comment (TheNeuralBit) [inline sdks/python/apache_beam/runners/interactive/pipeline_instrument.py:891]: Do you still need this default value for cacheable in case its not set elsewhere?

File: sdks/python/setup.py
Comment (TheNeuralBit) [inline sdks/python/setup.py:137]: I don't see anything at https://pypi.org/project/dataclasses/ explaining why we don't need a version bound, could you clarify?

It could make sense to add a `python_version<3.7`, so that if/when we drop python 3.6 support it's clear we can drop this.

**Review by TheNeuralBit — APPROVED** [review]
LGTM overall, just a couple minor comments/suggestions

File: sdks/python/apache_beam/runners/interactive/pipeline_instrument.py
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/pipeline_instrument.py:891]: No, this was in use back when Cacheable is not a class but a primitive dictionary. It is not needed anymore.

File: sdks/python/setup.py
Comment (KevinGG) [reply sdks/python/setup.py:137]: Yes it makes sense. I'll add a `python_version<3.7` TODO to remove it when we drop Python 3.6.

**More context on why no version bound**:

Based on https://pypi.org/project/dataclasses/

>this repo should only be used for historical purposes

If you look at the "Release History" section of it, the newer version of the package is for backport of Python 3.6.
For Python 3.7, the compatible version is actually lower (though installation is not needed at all).

Tested as something like: 

| Python version | Compatible dataclasses version |
| --------------- | -------------------------------- |
|3.7 | 0.5|
|3.8 | 0.6|

A range of versions doesn't make sense since each Python version uses a specific version of dataclasses.

File: sdks/python/apache_beam/runners/interactive/caching/cacheable.py
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/cacheable.py:48]: Thanks, the `pipeline_instrument` module still uses `Cacheable` as a mutable class. So we have to keep the explicit hash.

**Comment by KevinGG** [comment]
Force pushed to resolve a merge conflict.

File: sdks/python/setup.py
Comment (TheNeuralBit) [reply sdks/python/setup.py:137]: What I mean was to put a `python_version<3.7` on the requirement, that way it will just be ignored entirely for newer versions (and it will also remind us to drop the dep in the future).

Comment (TheNeuralBit) [reply sdks/python/setup.py:137]: It looks like we don't actually have any examples of this in our setup.py to draw from anymore! We used to do it all over the place because of py2.7 and 3.5. Here is an example of what I'm talking about: https://stackoverflow.com/a/32643122/2952954

Comment (TheNeuralBit) [reply sdks/python/setup.py:137]:
```suggestion
    'dataclasses;python_version<"3.7"',
```
WDYT about making it explicit like this?

Comment (KevinGG) [reply sdks/python/setup.py:137]: OMG, that's exactly what I wanted! Thanks, I'll make the change.

**Comment by tysonjh** [comment]
Anything holding this back from being merged?

**Comment by TheNeuralBit** [comment]
Just https://github.com/apache/beam/pull/14176#discussion_r592829797. I went ahead and applied the commit and Ill merge when CI is green.

**Comment by KevinGG** [comment]
> Just [#14176 (comment)](https://github.com/apache/beam/pull/14176#discussion_r592829797). I went ahead and applied the commit and Ill merge when CI is green.

Thanks @TheNeuralBit! I had a network outage yesterday right before I can commit the change

**Pull request merged**
